### PR TITLE
[SPARK-28876][SQL] fallBackToHdfs should not support Hive partitioned table

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR makes `spark.sql.statistics.fallBackToHdfs` not support Hive partitioned tables.

### Why are the changes needed?

The current implementation is incorrect for external partitions and it is expensive to support partitioned table with external partitions.

### Does this PR introduce any user-facing change?
Yes.  But I think it will not change the join strategy because partitioned table usually very large.

### How was this patch tested?
unit test

### CODE REVIEW

**Comment by wangyum** [comment]
cc @cloud-fan

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala [hunk 1]
```diff
@@ -114,22 +114,22 @@ class ResolveHiveSerdeTable(session: SparkSession) extends Rule[LogicalPlan] {
...
 class DetermineTableStats(session: SparkSession) extends Rule[LogicalPlan] {
   override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
-    case relation: HiveTableRelation
-        if DDLUtils.isHiveTable(relation.tableMeta) && relation.tableMeta.stats.isEmpty =>
-      val table = relation.tableMeta
-      val sizeInBytes = if (session.sessionState.conf.fallBackToHdfsForStatsEnabled) {
+    case relation @ HiveTableRelation(table, _, partitionCols)
+      if DDLUtils.isHiveTable(table) && table.stats.isEmpty =>
+      val conf = session.sessionState.conf
+      val sizeInBytes = if (conf.fallBackToHdfsForStatsEnabled && partitionCols.isEmpty) {
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala]: can we add some comment to explain why we should skip partitioned tables?

**Comment by cloud-fan** [comment]
LGTM

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala (hunk 1 above)
Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala]: Done

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 2]
```diff
@@ -1484,4 +1484,27 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
...
+        HiveUtils.CONVERT_METASTORE_PARQUET.key -> "false") {
+        withTempDir { tempDir =>
+          withTable("spark_28876") {
+            sql(s"""
+               |CREATE EXTERNAL TABLE spark_28876(id bigint)
+               |PARTITIONED BY (ds STRING)
+               |STORED AS PARQUET
+               |LOCATION '${tempDir.toURI}'
+             """.stripMargin)
+            sql("INSERT INTO TABLE spark_28876 PARTITION (ds='p1') SELECT 1")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: can we use `ADD PARTITION` to test external partitions？

Comment (wangyum) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: Done

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala [hunk 3]
```diff
@@ -1484,4 +1484,41 @@ class StatisticsSuite extends StatisticsCollectionTestBase with TestHiveSingleto
...
+            withTempDir { partitionDir =>
+              spark.range(5).write.mode(SaveMode.Overwrite).parquet(partitionDir.getCanonicalPath)
+              sql(s"ALTER TABLE spark_28876 ADD PARTITION (ds='p1') LOCATION '$partitionDir'")
+
+              val table = getCatalogTable("spark_28876")
+              assert(table.stats.isEmpty)
+              val partitions = spark.sessionState.catalog.listPartitions(table.identifier)
+              assert(partitions.size === 1)
+              partitions.foreach { p =>
+                // Partition directory is outside of the table directory
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: we can simplify it a little
```
withTempDir { dir =>
  val tableDir = new File(dir, "table")
  val partitionDir = new File(dir, "partition")
}
```

Then we can guarantee that partition directory is outside of the table directory

Comment (wangyum) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/StatisticsSuite.scala]: Done

**Comment by cloud-fan** [comment]
thanks, merging to master

**Pull request closed without merging**